### PR TITLE
Simplify planning of MERGE

### PR DESCRIPTION
**Pull request by martint**
Currently, the planner inserts a projection of the following shape to assemble the merged row:

    merge_row := (
        CASE
            WHEN ... THEN
                ROW(..., $not((present IS NULL)), <operation>, 0)
            WHEN ... THEN
                ROW(..., $not((present IS NULL)), <operation>, 1)
            ...
            ELSE
                ROW(<nulls>, $not((present IS NULL)), -1, -1)
        END)

This change replaces the ELSE branch to return a single null instead of a synthetic value with nulls. By reducing the size of the projection, it allows for wider tables to be used with MERGE.

Related to https://github.com/trinodb/trino/issues/15848

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
## Section
* TBD. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/test/java/io/trino/sql/planner/TestDeleteAndInsertMergeProcessor.java [hunk 1]
```diff
@@ -33,57 +32,17 @@
...
 import static io.trino.spi.type.BigintType.BIGINT;
 import static io.trino.spi.type.IntegerType.INTEGER;
-import static io.trino.spi.type.TinyintType.TINYINT;
 import static io.trino.spi.type.VarbinaryType.VARBINARY;
 import static io.trino.spi.type.VarcharType.VARCHAR;
 import static org.assertj.core.api.Assertions.assertThat;
 
 public class TestDeleteAndInsertMergeProcessor
 {
-    @Test
```
Comment (djsstarburst) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestDeleteAndInsertMergeProcessor.java:49]: Why did you decide to remove this test?

**Review by djsstarburst — APPROVED** [review]
This looked good to me and it looks like all the CI tests pass.  I wondered why the first test in `TestDeleteAndInsertMergeProcessor` was removed.

File: core/trino-main/src/test/java/io/trino/sql/planner/TestDeleteAndInsertMergeProcessor.java (hunk 1 above)
Comment (martint) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestDeleteAndInsertMergeProcessor.java:49]: The test makes very strong assumptions about how the blocks are assembled and is not set up to support a null in place of the merge row block. I just wanted to get the CI going while I figure out how to rewrite it.

**Comment by dougb** [comment]
Hi, I tried this PR and it seems to help with one of the tables I was having trouble MERGING, thank you!
Will do some more testing tomorrow.

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 2]
```diff
@@ -861,17 +862,16 @@ public MergeWriterNode plan(Merge merge)
...
-        rowBuilder.add(new Constant(TINYINT, -1L));
-        // The case number
-        rowBuilder.add(new Constant(INTEGER, -1L));
-
-        Case caseExpression = new Case(whenClauses.build(), new Row(rowBuilder.build()));
+        Case caseExpression = new Case(
+                whenClauses.build(),
+                new Constant(
+                        RowType.anonymous(ImmutableList.<Type>builder()
+                        .addAll(dataColumnSchemas.stream().map(ColumnSchema::getType).collect(Collectors.toList()))
```
Comment (hashhar) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:869]: nit: static import `toList`

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 3]
```diff
@@ -861,17 +862,16 @@ public MergeWriterNode plan(Merge merge)
...
+                        .add(BOOLEAN)
+                        .add(TINYINT)
+                        .add(INTEGER)
```
Comment (hashhar) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:872]: The only reason we have to keep the fields here is to make the type of the projection of the CASE expression match for each branch? They are not used in any way for the ELSE branch, are they?

**Review by hashhar — COMMENTED** [review]
Looks good, I only have questions for my own understanding.

We used to use the synthetic value as a "marker" to signal what operation to do.
Now we rely on the fact that the table scan will not produce a tuple with the value `CAST(null AS row(columns..., c1 boolean, c2 tinyint, c3 integer))` ever? Why is this assumption true? Is it true because there will never be a column in the table whose type matches the row we have created?

Or do I misunderstand something?

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java (hunk 3 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:872]: Exactly.

**Comment by martint** [comment]
The columns from the table scan are wrapped in a ROW(...) with some additional fields by a synthetic projection in the plan that handles the branches of the MERGE statement. The row is then consumed by the merge processor to decide if a row is being inserted or deleted. In one of the cases, we were producing a row that was meant to be ignored by the merge processor (i.e., when none of the MERGE branches matched).

**Review by hashhar — APPROVED** [review]

**Pull request merged**